File: ai_engine.py
"""AI pipeline on Claude/Groq API: JD understanding -> CV understanding -> Evaluation.

Supports Anthropic (Claude) and Groq (Llama) models.
"""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Any, Callable

import anthropic
import groq

ProgressFn = Callable[[str], None]

from prompts import (
    CV_EXTRACTION_PROMPT,
    EVALUATION_PROMPT,
    JD_EXTRACTION_PROMPT,
)

DEFAULT_MODEL = os.getenv("CV_SCORING_MODEL", "llama-3.3-70b-versatile")

SYSTEM_PROMPT = (
    "You are a panel of senior hiring experts: a Senior Hiring Manager (10+ yrs), "
    "a Senior Technical Recruiter, and a Senior CV Coach. "
    "You reason semantically like a real recruiter, never do keyword matching, "
    "and you ALWAYS return strict valid JSON only — no prose, no markdown fences."
)


@dataclass
class ScreeningResult:
    jd_understanding: dict[str, Any]
    cv_understanding: dict[str, Any]
    evaluation: dict[str, Any]


def _chat_json(
    api_key: str | None,
    model: str,
    user_prompt: str,
    max_tokens: int = 16000,
    progress: ProgressFn | None = None,
) -> dict:
    """Stream a request (Anthropic or Groq) and parse the full JSON response."""
    is_groq = "llama" in model.lower() or "mixtral" in model.lower() or "gemma" in model.lower()

    if is_groq:
        key = api_key or os.getenv("GROQ_API_KEY")
        if not key:
            raise RuntimeError("GROQ_API_KEY not set. Put it in .env or pass it via UI.")
        
        client = groq.Groq(api_key=key)
        safe_max_tokens = min(max_tokens, 8000) # Giới hạn max output của Groq
        
        stream = client.chat.completions.create(
            model=model,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": user_prompt}
            ],
            max_tokens=safe_max_tokens,
            temperature=0.0,
            stream=True
        )
        
        text_chunks = []
        for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                text_chunks.append(chunk.choices[0].delta.content)
                
        text = "".join(text_chunks).strip()
        if progress:
            progress("   ↳ (Groq stream completed fast)")
        return _parse_json(text)

    else:
        key = api_key or os.getenv("ANTHROPIC_API_KEY")
        if not key:
            raise RuntimeError("ANTHROPIC_API_KEY not set. Put it in .env or pass it via UI.")
        
        client = anthropic.Anthropic(api_key=key)
        with client.messages.stream(
            model=model,
            max_tokens=max_tokens,
            system=SYSTEM_PROMPT,
            messages=[{"role": "user", "content": user_prompt}],
        ) as stream:
            message = stream.get_final_message()

        text = "".join(b.text for b in message.content if b.type == "text").strip()
        if progress:
            usage = getattr(message, "usage", None)
            if usage is not None:
                progress(f"   ↳ tokens: in {usage.input_tokens}, out {usage.output_tokens}")
        return _parse_json(text)


def _parse_json(text: str) -> dict:
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        cleaned = text.strip()
        if cleaned.startswith("```"):
            cleaned = cleaned.strip("`")
            if cleaned.lower().startswith("json"):
                cleaned = cleaned[4:]
        start = cleaned.find("{")
        end = cleaned.rfind("}")
        if start != -1 and end != -1:
            cleaned = cleaned[start : end + 1]
        return json.loads(cleaned)


def extract_jd(
    api_key: str | None,
    jd_text: str,
    model: str = DEFAULT_MODEL,
    progress: ProgressFn | None = None,
) -> dict:
    prompt = JD_EXTRACTION_PROMPT.replace("{jd}", jd_text)
    return _chat_json(api_key, model, prompt, progress=progress)


def extract_cv(
    api_key: str | None,
    cv_text: str,
    model: str = DEFAULT_MODEL,
    progress: ProgressFn | None = None,
) -> dict:
    prompt = CV_EXTRACTION_PROMPT.replace("{cv}", cv_text)
    return _chat_json(api_key, model, prompt, progress=progress)


def evaluate(
    api_key: str | None,
    jd_json: dict,
    cv_json: dict,
    cv_raw: str,
    model: str = DEFAULT_MODEL,
    progress: ProgressFn | None = None,
) -> dict:
    prompt = EVALUATION_PROMPT.format(
        jd_json=json.dumps(jd_json, ensure_ascii=False, indent=2),
        cv_json=json.dumps(cv_json, ensure_ascii=False, indent=2),
        cv_raw=cv_raw[:8000],
    )
    result = _chat_json(api_key, model, prompt, max_tokens=16000, progress=progress)
    return _normalize_evaluation(result)


def screen_candidate(
    jd_text: str,
    cv_text: str,
    api_key: str | None = None,
    model: str = DEFAULT_MODEL,
    progress: ProgressFn | None = None,
) -> ScreeningResult:
    def log(msg: str) -> None:
        if progress:
            progress(msg)

    provider = "Groq" if ("llama" in model.lower() or "mixtral" in model.lower()) else "Claude"
    log(f"🔌 Connecting to {provider} ({model})...")

    log("📋 Step 1/3 — Reading Job Description (extracting required skills, seniority, hidden expectations)")
    jd_json = extract_jd(api_key, jd_text, model=model, progress=progress)
    log(f"   ✓ JD parsed — role: {jd_json.get('role_title','?')} · seniority: {jd_json.get('seniority_level','?')}")

    log("📄 Step 2/3 — Reading CV (extracting profile, trajectory, bullets, red-flag signals)")
    cv_json = extract_cv(api_key, cv_text, model=model, progress=progress)
    log(
        f"   ✓ CV parsed — candidate: {cv_json.get('candidate_name','?')} · "
        f"trajectory: {cv_json.get('career_trajectory','?')[:60]}"
    )

    log("🧠 Step 3/3 — Senior panel evaluating (scoring, rewrites, hiring decision)")
    evaluation = evaluate(api_key, jd_json, cv_json, cv_text, model=model, progress=progress)
    log(
        f"   ✓ Evaluation done — score: {evaluation.get('overall_score','?')}/100 · "
        f"grade: {evaluation.get('grade','?')} · "
        f"decision: {evaluation.get('hiring_decision',{}).get('recommendation','?')}"
    )

    return ScreeningResult(
        jd_understanding=jd_json,
        cv_understanding=cv_json,
        evaluation=evaluation,
    )


def _normalize_evaluation(ev: dict) -> dict:
    """Enforce schema shape so the UI never crashes on missing keys."""
    ev.setdefault("overall_score", 0)
    ev.setdefault("grade", "Weak")
    ev.setdefault("confidence", 0.0)
    ev.setdefault("summary", "")
    ev.setdefault("strengths", [])
    ev.setdefault("weaknesses", [])
    dims = ev.setdefault("dimension_scores", {})
    for k in ("jd_match", "cv_quality", "experience_depth", "formatting", "risk"):
        dims.setdefault(k, 0)
    imp = ev.setdefault("improvements", {})
    imp.setdefault("content_issues", [])
    imp.setdefault(
        "skill_gaps",
        {"critical_missing": [], "secondary_missing": [], "transferable": []},
    )
    imp.setdefault("positioning_issues", [])
    imp.setdefault("experience_issues", [])
    imp.setdefault("formatting_issues", [])
    imp.setdefault("red_flags", [])
    sug = ev.setdefault("suggestions", {})
    sug.setdefault("micro_fixes", [])
    sug.setdefault("macro_fixes", [])
    sug.setdefault("strategic_advice", [])
    ev.setdefault(
        "hiring_decision",
        {"recommendation": "Consider", "reason": "", "top_risks": []},
    )
    return ev

File: lib/ai-engine.ts
// lib/ai-engine.ts  — server-side only
import Anthropic from "@anthropic-ai/sdk";
import Groq from "groq-sdk";
import {
  Evaluation,
  JDUnderstanding,
  CVUnderstanding,
  ScreeningResult,
} from "./types";
import {
  SYSTEM_PROMPT,
  JD_EXTRACTION_PROMPT,
  CV_EXTRACTION_PROMPT,
  EVALUATION_PROMPT,
} from "./prompts";

export const DEFAULT_MODEL = process.env.CV_SCORING_MODEL ?? "claude-sonnet-4-6";

export function isGroqModel(model: string) {
  return ["llama", "mixtral", "gemma", "openai/gpt-oss"].some((k) => model.toLowerCase().includes(k));
}

function parseJSON<T>(raw: string): T {
  let text = raw.trim();
  if (text.startsWith("```")) {
    text = text.replace(/^```[a-z]*\n?/, "").replace(/```$/, "").trim();
  }
  const start = text.indexOf("{");
  const end = text.lastIndexOf("}");
  if (start !== -1 && end !== -1) text = text.slice(start, end + 1);
  return JSON.parse(text) as T;
}

let currentGroqKeyIndex = 0;

async function callGroq<T>(model: string, prompt: string, maxTokens = 8000): Promise<T> {
  const keysStr = process.env.GROQ_API_KEY;
  if (!keysStr) throw new Error("GROQ_API_KEY is not set in .env.local");

  const keys = keysStr.split(',').map(k => k.trim()).filter(Boolean);
  if (keys.length === 0) throw new Error("No valid GROQ_API_KEY found");

  let lastError: any;

  for (let attempt = 0; attempt < keys.length; attempt++) {
    const key = keys[currentGroqKeyIndex % keys.length];
    currentGroqKeyIndex++;

    try {
      const client = new Groq({ apiKey: key });
      const stream = await client.chat.completions.create({
        model,
        messages: [
          { role: "system", content: SYSTEM_PROMPT },
          { role: "user", content: prompt },
        ],
        max_tokens: Math.min(maxTokens, 8000), // Groq output limit
        temperature: 0,
        stream: true,
      });

      const chunks: string[] = [];
      for await (const chunk of stream) {
        const delta = chunk.choices[0]?.delta?.content;
        if (delta) chunks.push(delta);
      }
      return parseJSON<T>(chunks.join(""));
    } catch (error: any) {
      lastError = error;
      if (error?.status === 429 || error?.response?.status === 429 || error?.message?.includes("429")) {
        console.warn(`[Groq] Key rate limited, switching to next key. Attempt ${attempt + 1}/${keys.length}`);
        continue;
      }
      throw error;
    }
  }

  throw lastError;
}

async function callClaude<T>(model: string, prompt: string, maxTokens = 8000): Promise<T> {
  const key = process.env.ANTHROPIC_API_KEY;
  model = process.env.CLAUDE_MODEL || "none";
  if (!key) throw new Error("ANTHROPIC_API_KEY is not set in .env.local");

  const client = new Anthropic({ apiKey: key });
  const stream = await client.messages.stream({
    model,
    max_tokens: maxTokens,
    system: SYSTEM_PROMPT,
    messages: [{ role: "user", content: prompt }],
  });
  const msg = await stream.finalMessage();
  const text = msg.content
    .filter((b): b is Anthropic.TextBlock => b.type === "text")
    .map((b) => b.text)
    .join("");

  return parseJSON<T>(text);
}

export type Lang = "en" | "vi";

function langDirective(lang: Lang): string {
  if (lang === "vi") {
    return (
      "\n\nLANGUAGE REQUIREMENT: Write ALL natural-language string values in the output JSON " +
      "in Vietnamese (Tiếng Việt). This includes: summary, strengths, weaknesses, reason, " +
      "problem descriptions, improved_version, rewritten_summary, all suggestion items, " +
      "red_flag descriptions, and any human-readable text. Keep JSON KEYS, enum values " +
      "(grade: Strong Hire/Good Fit/Moderate/Weak, recommendation: Hire/Consider/Reject, " +
      "issue_type), numbers, and technical terms (programming languages, frameworks, tools) " +
      "in English. Do NOT translate proper nouns.\n"
    );
  }
  return "\n\nLANGUAGE REQUIREMENT: Write all natural-language output in English.\n";
}

async function callLLM<T>(model: string, prompt: string, lang: Lang, maxTokens = 8000): Promise<T> {
  const finalPrompt = prompt + langDirective(lang);
  return isGroqModel(model)
    ? callGroq<T>(model, finalPrompt, maxTokens)
    : callClaude<T>(model, finalPrompt, maxTokens);
}

export async function extractJD(model: string, jdText: string, lang: Lang = "en"): Promise<JDUnderstanding> {
  return callLLM<JDUnderstanding>(model, JD_EXTRACTION_PROMPT(jdText), lang);
}

export async function extractCV(model: string, cvText: string, lang: Lang = "en"): Promise<CVUnderstanding> {
  return callLLM<CVUnderstanding>(model, CV_EXTRACTION_PROMPT(cvText), lang);
}

export async function evaluate(
  model: string,
  jdJson: JDUnderstanding,
  cvJson: CVUnderstanding,
  cvRaw: string,
  position: string,
  level: string,
  compareMarket: boolean,
  lang: Lang = "en"
): Promise<Evaluation> {
  const ev = await callLLM<Evaluation>(
    model,
    EVALUATION_PROMPT(
      JSON.stringify(jdJson, null, 2),
      JSON.stringify(cvJson, null, 2),
      cvRaw,
      position,
      level,
      compareMarket
    ),
    lang as Lang,
    isGroqModel(model) ? 8000 : 16000
  );
  return normalizeEvaluation(ev, compareMarket);
}

export async function screenCandidate(
  jdText: string,
  cvText: string,
  model: string,
  position: string,
  level: string,
  compareMarket: boolean,
  onProgress: (msg: string) => void,
  lang: Lang = "en"
): Promise<ScreeningResult> {
  onProgress("📋 Step 1/3 — Reading Job Description…");
  const jd = await extractJD(model, jdText, lang);
  onProgress(`   ✓ JD parsed — ${jd.role_title} · ${jd.seniority_level}`);

  onProgress("📄 Step 2/3 — Parsing CV…");
  const cv = await extractCV(model, cvText, lang);
  onProgress(`   ✓ CV parsed — ${cv.candidate_name} · ${cv.career_trajectory.slice(0, 60)}`);

  onProgress("🧠 Step 3/3 — Senior panel evaluating…");
  const evaluation = await evaluate(model, jd, cv, cvText, position, level, compareMarket, lang);
  onProgress(
    `   ✓ Done — ${evaluation.overall_score}/100 · ${evaluation.grade} · ${evaluation.hiring_decision.recommendation}`
  );

  if (compareMarket && evaluation.market_comparison) {
    onProgress(`📊 Market insight — Top ${evaluation.market_comparison.market_rank_top_pct}% of ${position} (${level})`);
  }

  return { jd_understanding: jd, cv_understanding: cv, evaluation };
}

function normalizeEvaluation(ev: Partial<Evaluation>, compareMarket: boolean): Evaluation {
  const base: Evaluation = {
    overall_score: 0,
    grade: "Weak",
    confidence: 0,
    summary: "",
    strengths: [],
    weaknesses: [],
    dimension_scores: {
      jd_match: 0,
      cv_quality: 0,
      experience_depth: 0,
      formatting: 0,
      risk: 0,
    },
    improvements: {
      content_issues: [],
      skill_gaps: { critical_missing: [], secondary_missing: [], transferable: [] },
      positioning_issues: [],
      experience_issues: [],
      formatting_issues: [],
      red_flags: [],
    },
    suggestions: { micro_fixes: [], macro_fixes: [], strategic_advice: [] },
    hiring_decision: { recommendation: "Consider", reason: "", top_risks: [] },
  };

  const result = { ...base, ...ev } as Evaluation;

  // Normalize market_comparison if it was requested but came back incomplete
  if (compareMarket && result.market_comparison) {
    result.market_comparison = {
      market_rank_top_pct: result.market_comparison.market_rank_top_pct ?? 50,
      market_context: result.market_comparison.market_context ?? "",
      market_trends: result.market_comparison.market_trends ?? [],
      skill_alignment: result.market_comparison.skill_alignment ?? [],
      improvement_priority: result.market_comparison.improvement_priority ?? [],
      smart_action_plan: result.market_comparison.smart_action_plan ?? [],
    };
  }

  return result;
}
